### DISCUSSION TITLE
Coordinator 패턴 적용 방법

### DISCUSSION THREAD

**Discussion by nomatterjun** [Favor-Gift-Reminder/Favor-iOS#7] [Show and tell]
## 🙋 논제

Coordinator에서  ViewController에 reactor를 주입할 수 있도록 생성자를 만들어 구현하였습니다.
이 방법이 괜찮은지 아래 커밋 기록 한번 리뷰 부탁드려요! 

8146958c227973989b10c98cd6315b65ccbec13d

https://github.com/Favor-Gift-Reminder/Favor-iOS/blob/66f436c90974923f1acdefa60596b304c7abec07/Favor/Favor/Sources/Scenes/Home/ViewControllers/HomeVC.swift#L22-L29

https://github.com/Favor-Gift-Reminder/Favor-iOS/blob/66f436c90974923f1acdefa60596b304c7abec07/Favor/Favor/Sources/Scenes/Home/Coordinator/HomeCoordinator.swift#L18-L22

### 해당 커밋 구현 내역
- AppCoordinator를 Root Coordinator로서 Scene Delegate에서 푸쉬
- 비어있는 임시 HomeVC / HomeReactor / HomeCoordinator 생성하여 푸쉬
- Coordinator의 `start()` 메서드에서 필요한 데이터들을 주입하면서 초기화(`Init`)

**Comment by EnnzoKim** [comment]
+  개인적으로 ViewController에 생성자를 추가하는 것을 별로 선호하지 않아서.. 
왜냐하면 아래 코드 같이 기본 생성자를 쓰지 않을 경우엔 `required`라서 조금 지저분해지는 느낌?이 들어!
~~물론 생성자가 필연적인 경우는 사용해야겠지만...~~
그래서 개인적으로는 공식 ReactorKit Readme에서 권장하고 있는 방식처럼 하면 어떨까 라는 의견이야
```swift 
homeViewController.reactor = homeReactor
```

+ 하지만 반대로 `reactor`에 `coordinator`를 생성자를 넣는건 반대로 찬성이야! 
이렇게 되면 `BaseViewController`에 있는 `coordinator` 프로퍼티가 필요하지 않을 것 같아.
혹시 이렇게 한 의도가  state에 변화 감지에 따른 화면 전환을 coordinator에서 하지 않고 reactor에서 처리해주기 위해서 이렇게 구현한걸까?

+ 단점은 모든 reactor에 명시적으로 `weak var coordinator: SomeCoordinator?`와 coordinator에 대한 생성자를 모두 적어줘야 한다는 것?이 있을 것 같아

읽어보고 의견남겨줘!

**Reply by nomatterjun** [reply]
- 나는 의존성을 주입이 필요하면 생성자가 호출될 때 모두 주입되고 시작하는 걸 선호해서 뷰컨에도 굳이 생성자를 넣어준 거였어. 
``` Swift
homeViewController.reactor = homeReactor
```
굳이굳이 BaseVewController에 생성자를 넣어주고 할 바에는 위 한 줄로 끝내는 게 맞는 것 같아! 이건 아마 내 평소 코딩 습관에 따라서 어려운 길로 돌아간 느낌이야. 이 부분은 수정해서 커밋해둘게!
- `reactor`에 `coordinator`를 넣는 건 내가 생각했을 땐 우리가 mvvm 구조를 채택한 이상 뷰컨에서 일어나는 모든 이벤트들이 reactor에 한 번은 거쳐가야 한다고 생각해서야. 유닛 테스트를 진행할 때 코디네이터가 뷰컨에 주입되어 있으면 버튼 탭과 같은 `Action`에 따른 `Mutate`를 테스트해볼 때 실제로 화면이 이동되었는지와 같은 결과는 확인을 할 수 없을 것 같아.
- 또 Coordinator 패턴을 사용하는 이유 중에 의존성 주입을 코디네이터가 전담하게 하려는 목적도 있잖아? 그러면 
``` Swift
self.coordinator?.start(childCoordinator: someCoordinator, with: didAcceptRequest)
```
위처럼 뷰에 데이터를 함께 전달해주어야 할 때는 뷰컨이 데이터 모델(도메인 레이어)에까지 접근해야하는 일이 일어나지 않을까 싶어. 그래서 Clean하지 못한 설계가 되지 않을까 하는 생각이야!

**Reply by EnnzoKim** [reply]
오케이~ reactor에서 coordinator 받는건 좋은 것 같아
나는 이전에 coordinator에서 직접적으로 state에 이벤트들을 받았는데
reactor에서 받는게 여러모로 좋을 것 같아!

**Comment by nomatterjun** [comment]
논의 내용 da1efae5a253138e5a76d87c47f086e53bfbb822 에서 적용 완료! 👍👍
